### ISSUE TITLE
[2.1] & [3.0]: BBC - hr within table breaks table

### ISSUE THREAD

**Issue by sbulen** [SimpleMachines/SMF#8299]
### Basic Information

Using an [hr] tag within a bbc table can break table formatting, cause it to display the bbc code instead of acting upon it...

This is true in message bodies, in signatures, and in signature & post previews as well.  

### Steps to reproduce

1.  Put this in a post or signature:
```
[table]
[tr][td][b][color=#0066cc]Slava[/color]
[color=#ffcc00][hr]Ukraini![/color][/b][/td][/tr]
[tr][td][/td][/tr]
[tr][td][sub]"Therefore, you should only allow people that you trust, implicitly, to have such access." -Douglas[/sub][/td][/tr]
[/table]

```

### Expected result

Properly formatted table, with an hr that produces a line within the cell.  

### Actual result

![image](https://github.com/user-attachments/assets/dc294a13-c7d8-4105-8e83-54d0310b5887)

### Version/Git revision

3.0 Alpha 2 & 2.1.4

### Database Engine

All

### Database Version

8..4 MySQL

### PHP Version

8.3.8

### Additional Information

Related forum thread:
https://www.simplemachines.org/community/index.php?topic=588511.0

**Comment by jdarwood007** [comment]
This isn't a bug.  `[hr]` is considered a block level element.  While `[b]` and `[color` are not and considered inline elements.

Proper BBC for this would be:
```
[table]
[tr][td][b][color=#0066cc]Slava[/color][/b]
[hr]
[b][color=#ffcc00]Ukraini![/color][/b][/td][/tr]
[tr][td][/td][/tr]
[tr][td][sub]"Therefore, you should only allow people that you trust, implicitly, to have such access." -Douglas[/sub][/td][/tr]
[/table]
```

You can also make that invalid tag work in browsers.  From what I can read on a quick google search, hr is considered a `display: block;` and should thus be a block level element in HTML output.

If you believe I'm wrong here, reopen this to discuss.

**Comment by sbulen** [comment]
This is incorrect....  Lex & live627 both correctly pointed out that hr is valid within an html table and will only do the width of the cell. 

Please read the initial discussion.

**Comment by Sesquipedalian** [comment]
Reopening for now, with further discussion pending.

**Comment by jdarwood007** [comment]
If we decide that hr is not a block level the fix is simple:
```
			array(
				'tag' => 'hr',
				'type' => 'closed',
				'content' => '<hr>',
				'block_level' => true,
			),
```

Change to false and it works.

But I will point out that the BBC code will generate invalid HTML.  You have a [hr] inside of a [b] and [color].  Which are converted to HTML span tag.  You are then putting the hr inside the span.  You end up with this effectively:

```
<b><span style="{color}">Text<hr>More</span></b>
```

The relevant portion of the BBC you provided:
```
[b][color=#0066cc]Slava[/color]
[color=#ffcc00][hr]Ukraini![/color][/b]
```

I provided an example BBC that corrects this by closing the bold, doing the hr and then opening the bold and color

**Comment by sbulen** [comment]
The question is whether we should model what html would do in this instance.  In HTML it works:
```
<table>
<tr><td>R1C1</td><td>R1C2</td><td>R1C3</td></tr>
<tr><td>R2C1</td><td>R2C2 - TOP<hr>R2C2 - BOTTOM</td><td>R2C3</td></tr>
<tr><td>R3C1</td><td>R3C2</td><td>R3C3</td></tr>
</table>
```
Displays as:
![Image](https://github.com/user-attachments/assets/1226bce9-e72b-4c66-9195-2d3cd8d04189)

Lots of discussions out there on hr usage within tables, most folks say it's ill-advised, but the Mozilla docs say hd is valid within a table, so it's _technically_ supported(?).  

IMO, if it's tough, just don't support it.  I don't think that was a popular view in the support forum though.  Folks thought it should work.  

I just logged the issue in my likely misguided Sisyphean quest to keep the GH issue list & the Bug Report board in sync...  If we're going to close it out, we should close it out on the forum, too.

**Comment by jdarwood007** [comment]
The BBC equivalent of the HTML you provided is:

```
[table]
[tr][td]R1C1[/td][td]R1C2[/td][td]R1C3[/td][/tr]
[tr][td]R2C1[/td][td]R2C2 - TOP[hr]R2C2 - BOTTOM[/td][td]R2C3[/td][/tr]
[tr][td]R3C1[/td][td]R3C2[/td][td]R3C3[/td][/tr]
[/table]
```
Which works in

SMF 3.0
<img width="464" alt="Image" src="https://github.com/user-attachments/assets/6b4becd3-c17a-4398-b2b4-ad6bb38e9ea7" />

SMF 2.1
<img width="493" alt="Image" src="https://github.com/user-attachments/assets/d0ea8a55-c715-4478-a8ee-75c424d1cc65" />

SMF 2.0
<img width="522" alt="Image" src="https://github.com/user-attachments/assets/11a99b6b-be88-4951-8f7f-a48bf313afae" />

The sample you provided initially is invalid BB Code, because the HR BB Code is used inside the color BB Code tag..  If you close out the color and bold BB Code tags and then use the HR BB Code, then reopen your bold and color BBC.  It works as you expect.

Your initial sample is broken and will produce the error you see, because it was used wrong.

If this needs "fixed", I provided information as to what it would take to allow the invalid HTML to be produced.
If this is being produced by SCEditor, then we do have a bug, but only that the editor should do as I mentioned and close out the tags properly before doing the HR and then reopen them.

**Comment by sbulen** [comment]
I think you are correct.  Move that [hr] to a logical place & it works.

I went back to the original post &, I think, there were lots of challenges with folks having copy & paste errors.

There is only one screenshot of one that wasn't working, and tbh, I don't think he hit the preview button, because it works fine...
(https://imgur.com/R7LPkNO)

**Comment by jdarwood007** [comment]
I tried that BBC in the preview and it worked.  Can you reproduce it?

```
[table][tr][td][hr]
Alex
[hr]
[/td]
[/tr]
[/table]
```

<img width="863" height="399" alt="Image" src="https://github.com/user-attachments/assets/a80047b3-8073-48ee-b4c7-3ab8a18397d5" />

**Comment by sbulen** [comment]
Nope.  The way to reproduce that screenshot is to make a correction (add a previously missing [table]) and not hit the preview button...

**Comment by jdarwood007** [comment]
Safe to close now?

If anyone can provide a valid BBC sample that produces the error or can get sceditor to generate the invalid BBC, we have a bug and we can reopen this and address it.

**Comment by sbulen** [comment]
I think so.  & I left a note on the original forum thread, which should probably be moved to fixed/bogus bugs.
